internal/encryption/method/external/command.go

```diff
@@ -0,0 +1,188 @@
+}
+
+func (c command) run(command []string, input any) ([]byte, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+	defer cancel()
+
+	inputData, err := json.Marshal(input)
+	if err != nil {
+		return nil, &method.ErrCryptoFailure{
+			Message: "failed to marshal input",
+			Cause:   err,
+		}
+	}
+
+	stderr := &bytes.Buffer{}
+
+	cmd := exec.CommandContext(ctx, command[0], command[1:]...) //nolint:gosec //Launching external commands here is the entire point.
+
+	handler := &ioHandler{
+		false,
+		bytes.NewBuffer(inputData),
+		[]byte{},
+		cancel,
+		nil,
+	}
+
+	cmd.Stdin = handler
+	cmd.Stdout = handler
+	cmd.Stderr = stderr
+	if err := cmd.Run(); err != nil {
+		if handler.err != nil {
+			return nil, &method.ErrCryptoFailure{
+				Message:          "external encryption method failure",
+				Cause:            handler.err,
+				SupplementalData: fmt.Sprintf("Stderr:\n-------\n%s\n", stderr.String()),
+			}
+		}
+
+		var exitErr *exec.ExitError
+		if errors.As(err, &exitErr) {
+			if exitErr.ExitCode() != 0 {
+				return nil, &method.ErrCryptoFailure{
+					Message:          "external encryption method exited with non-zero exit code",
+					Cause:            err,
+					SupplementalData: fmt.Sprintf("Stderr:\n-------\n%s\n", stderr.String()),
+				}
+			}
+		}
+		return nil, &method.ErrCryptoFailure{
+			Message:          "external encryption method exited with an error",
+			Cause:            err,
+			SupplementalData: fmt.Sprintf("Stderr:\n-------\n%s\n", stderr.String()),
+		}
+	}
+
+	var result *OutputV1
+	decoder := json.NewDecoder(bytes.NewReader(handler.output))
+	decoder.DisallowUnknownFields()
+	if err := decoder.Decode(&result); err != nil {
+		return nil, &method.ErrCryptoFailure{
+			Message:          "external encryption method returned an invalid JSON",
+			Cause:            err,
+			SupplementalData: fmt.Sprintf("Stderr:\n-------\n%s\n", stderr.String()),
+		}
+	}
+
+	return result.Payload, nil
+}
+
+type ioHandler struct {
+	headerFinished bool
+	input          *bytes.Buffer
+	output         []byte
+	cancel         func()
+	err            error
+}
+
+func (i *ioHandler) Write(p []byte) (int, error) {
+	i.output = append(i.output, p...)
+	n := len(p)
+	if i.headerFinished {
+		// Header is finished, just collect the output.
+		return n, nil
+	}
+	// Check if the full header is present.
+	parts := strings.SplitN(string(i.output), "\n", 2) //nolint:mnd //This rule is dumb.
+	if len(parts) == 1 {
+		return n, nil
+	}
+	var header Header
+	// Note: this is intentionally not using strict decoding. Later protocol versions may introduce additional header
+	// fields.
+	if jsonErr := json.Unmarshal([]byte(parts[0]), &header); jsonErr != nil {
+		err := fmt.Errorf("failed to unmarshal header from external binary (%w)", jsonErr)
+		i.err = err
+		i.cancel()
+		return n, err
+	}
+
+	if header.Magic != Magic {
+		err := fmt.Errorf("invalid magic received from external key provider: %s", header.Magic)
```

Comment:
Given that you accidentally copy pasted and forgot "key provider" here, would it make sense to make the ioHandler shared between the external kp and method?

Reply:
I'm thinking about abstracting out the command and buffer handling that's present in both areas.  It's up to you, I just ask anytime I've seen code that's been copied and not fully renamed.